AppDelegate.swift

```diff
@@ -0,0 +1,122 @@
+//
+//  AppDelegate.swift
+//  srushtika-tut-fixes
+//
+//  Created by Srushtika Neelakantam on 06/01/2020.
+//  Copyright © 2020 Srushtika Neelakantam. All rights reserved.
+//
+
+import UIKit
+import Ably
+import UserNotifications
+
+let authURL = ""
+let API_KEY = ""
+let myClientId = "abc"
+let ablyClientOptions = ARTClientOptions()
+
+
+@UIApplicationMain
+class AppDelegate: UIResponder, UIApplicationDelegate, ARTPushRegistererDelegate {
+    func didDeactivateAblyPush(_ error: ARTErrorInfo?) {
+        print("** push deactivated")
+    }
+    
+
+    var window: UIWindow?
+    var realtime: ARTRealtime!
+    var channel: ARTRealtimeChannel!
+    var subscribed = false
+    
+    //on launch, init ably and call activate
+    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
+        // Override point for customization after application launch.
+        
+        print("** app launched")
+
+       UNUserNotificationCenter.current().requestAuthorization(options:[.badge, .alert, .sound]) { (granted, err) in
+           DispatchQueue.main.async() {
+               UIApplication.shared.registerForRemoteNotifications()
+               print("** request to show notifications successful")
+           }
+       }
+        
+       self.realtime = self.getAblyRealtime()
+       self.realtime.push.activate()
+        
+        return true
+    }
+    
+    //register with APNs - returns deviceToken
+    func application(_ application: UIApplication, didRegisterForRemoteNotificationsWithDeviceToken deviceToken: Data) {
+        DispatchQueue.main.async() {
```

Comment:
This is not necessary. The user doesn't need to call the `ARTPush.didRegisterForRemoteNotifications` in the main thread. The lib can handle it. I know that old tutorials is doing that but it was because the library wasn't thread safe and now it is. You can just do:

```
func application(_ application: UIApplication, didRegisterForRemoteNotificationsWithDeviceToken deviceToken: Data) {
    print("** didRegisterForRemoteNotificationsWithDeviceToken")
    ARTPush.didRegisterForRemoteNotifications(withDeviceToken: deviceToken, realtime: self.getAblyRealtime())
}
```

It's simple for the user.